# STAT 207 Lab Assignment 2 - Individual Part - [25 total points] - Solutions

## Dataframe Manipulation and Cleaning

Due: Tuesday, September 5, 11:59pm CST

<hr>


## <u>Case Study</u>: Steam Game Play and Purchase Analysis
Steam is the world's most popular PC Gaming hub. With a massive collection that includes everything from AAA blockbusters to small indie titles, great discovery tools can be super valuable for Steam. We would like to use the following dataset below to learn more about Steam player behavior.

The dataset we will be analyzing is comprised of a random sample of 500 steam users and their game play and purchase behaviors. It has the following columns:
* **user_id**
* **game_name**,
* **activity**:
    - purchase: indicating that the user has *purchased* the corresponding game
    - play: indicating that the user has *played* the corresponding game (for at least some amount of time.) 
* **hours_played_if_play**:
    - if the row corresponds to a 'play' activity, this number represents the number of hours the user has played the game
    - if the row corresponds to a 'purchase' activity, this number is always a 1 (and means nothing... it's a placeholder).

<hr>


### <u>Research Goals</u>

In this analysis, we will consider a game's **"all-star" player** to be the person who had played this game for the most amount of hours.

1. **Missing Value Cleaning**: Identify any implicit or explicit missing values in the dataset and drop any rows that have a missing value.
2. **Game Purchase vs. Play Behaviors**: What percent of games that were *bought* were actually *played* in this dataset?
3. **Top Game Buyer**: Learn more about the person who bought the most amount of games in this dataset. How many of these games did they actually play?
4. **"All Star Players"**: Who were the "all-star players" of the games? How long did they play this game?
5. **All Star Player Behavior**: What percent of games did that these "all star players" bought did they actually play?



### <u>Points </u>
<p style="margin-top:0in;margin-right:0in;margin-bottom:0in;margin-left:0in;font-size:11.0pt;font-family:&quot;Calibri&quot;,sans-serif;line-height:normal;"></p>
<table style="border: none;border-collapse: collapse;width:102pt;">
    <tbody>
        <tr>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;width:51pt;">Problem</td>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-left:none;width:51pt;">Points</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">6</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">5</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">6</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">7.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">7.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">9</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
    </tbody>
</table>
<p></p>




<hr>

### <u>STAT107 Review</u>
Part of this individual assignment will involve a review of STAT107 code that you have seen before including the following. Click on the links below if you need a refresher.

1. <a href="https://discovery.cs.illinois.edu/learn/Basics-of-Data-Science-with-Python/DataFrames-with-Conditionals/">Row Filtering</a>
2. <a href="https://discovery.cs.illinois.edu/guides/Modifying-DataFrames/dataframe-groupby/">The groupby() and reset_index() Functions</a>
2. <a href="https://discovery.cs.illinois.edu/guides/Modifying-DataFrames/sorting-a-dataframe-with-pandas/">The sort_values() Function</a>


### <u>Tutorial you Will Need</u>: Utilizing the *.shape* attribute

The **.shape** attribute for a given dataframe returns a **tuple** of two values, the first value in the tuple is the number of rows, the second value is the number of columns.

In [2]:
import pandas as pd
demo_df=pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
demo_df

a  b
0  1  4
1  2  5
2  3  6

We can index a tuple with 0 and 1 indices to extract the number of rows and columns from the dataframe separately.

In [3]:
#Number of rows and columns
demo_df.shape

(3, 2)

In [4]:
#Number of rows
demo_df.shape[0]

3

In [5]:
#Number of columns
demo_df.shape[1]

2

<hr>

### Imports

In [6]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

### 1. Read the csv
First, read the **steam_sample_500.csv** file into a dataframe and display the first five rows of the dataframe.

In [7]:
df = pd.read_csv('steam_sample_500.csv')
df.head()

user_id    game_name  activity hours_played_if_play
0  218323237       Dota 2  purchase                    1
1  218323237       Dota 2      play                 14.8
2  218323237  Tomb Raider  purchase                    1
3  218323237  Tomb Raider      play                  8.8
4  218323237         TERA  purchase                    1

### 2. How many rows does the dataframe have? (Use Python code to show this).

In [8]:
df.shape[0]

8601

### 3. Data cleaning and data type checking

We plan to analyze the **"hours_played_if_play"** variable in this assignment. In order to ensure that our subsequent functions work appropriately on this numerical variable, you should do the following.
1. Identify any values (if any) that have been encoded in the csv to represent a **missing value**. 
2. Make sure that the values in your **"hours_played_if_play"** variable are represented as numerical values (either float64 or int64 object types). If not, fix this using the methods we discussed in class.
3. Delete any rows in your dataframe that have missing values (both explicit or implict). Remember, some "missing values", in the way that they're currently encoded in the csv may not be initially detectable by Python. It is your job to figure out what these implicit missing values are and make it so Python does detect them. And then delete any rows that have them.

In [9]:
df["hours_played_if_play"]

0          1
1       14.8
2          1
3        8.8
4          1
        ... 
8596       1
8597       1
8598     0.4
8599       1
8600     0.3
Name: hours_played_if_play, Length: 8601, dtype: object

In [10]:
df.dtypes

user_id                  int64
game_name               object
activity                object
hours_played_if_play    object
dtype: object

In [11]:
df["hours_played_if_play"].describe()

count     8601
unique     503
top          1
freq      5444
Name: hours_played_if_play, dtype: object

In [12]:
df["hours_played_if_play"].unique()

array(['1', '14.8', '8.8', '0.9', '1784', '626', '176', '69', '48',
       '16.1', '13.6', '11.7', '10.2', '5.8', '4.1', '2.3', '2.2', '0.6',
       '0.4', '0.3', '1118', '17.2', '5.9', '22', '86', 'data suppressed',
       '1.5', '1.2', '0.8', '290', '60', '14.1', '12.5', '9', '7.3',
       '4.8', '3.2', '3', '0.7', '0.5', '2.1', '191', '154', '100', '65',
       '64', '12.8', '9.7', '3.5', '2.6', '6.9', '1395', '25', '12.3',
       '10.6', '8.2', '6.3', '6', '5.4', '4', '1.8', '0.1', '110', '299',
       '21', '1.9', '258', '239', '45', '1.1', '10.7', '2.7', '30', '443',
       '23', '9.1', '2526', '1906', '725', '0.2', '2.4', '608', '337',
       '115', '95', '90', '88', '87', '74', '72', '70', '63', '58', '53',
       '52', '51', '47', '43', '37', '35', '34', '31', '29', '28', '27',
       '26', '24', '20', '19.6', '19', '17.8', '17', '15.8', '15', '13.8',
       '13.2', '12.1', '12', '10.5', '9.8', '9.4', '9.3', '8.4', '7.4',
       '7', '6.8', '6.4', '5.7', '5.1', '5', '4.6', '4.

In [13]:
df = pd.read_csv('steam_sample_500.csv'), na_values = ["data suppressed", "missing data"])

In [14]:
df.dtypes

user_id                   int64
game_name                object
activity                 object
hours_played_if_play    float64
dtype: object

In [15]:
df["hours_played_if_play"].mean()

22.008986148294728

### 4. Number of Dropped Rows 

How many rows with missing values did you delete from your original dataframe?

In [20]:
df.isna().sum()

user_id                  0
game_name                0
activity                 0
hours_played_if_play    10
dtype: int64

In [21]:
df.shape

(8601, 4)

In [22]:
df = df.dropna()

In [23]:
df.shape

(8591, 4)

### 5. Separate the dataframes. 

Next, in this analysis we would like to analyze the game purchase information and game play information separately. First create two new dataframes:
* one that is comprised of the purchase rows
* the other that is comprised of the play rows.

In [25]:
purchase_dataframe = pd.DataFrame(df[df["activity"] == "purchase"])
purchase_dataframe

user_id                                     game_name  activity  \
0     218323237                                        Dota 2  purchase   
2     218323237                                   Tomb Raider  purchase   
4     218323237                                          TERA  purchase   
6     126340495                                        Dota 2  purchase   
8     126340495               Counter-Strike Global Offensive  purchase   
...         ...                                           ...       ...   
8593    7519923                                Counter-Strike  purchase   
8595    7519923                 Counter-Strike Condition Zero  purchase   
8596    7519923  Counter-Strike Condition Zero Deleted Scenes  purchase   
8597  103963922                               Team Fortress 2  purchase   
8599  300991661                                        Dota 2  purchase   

      hours_played_if_play  
0                      1.0  
2                      1.0  
4                      1.0  
6                      1.0  
8                      1.0  
...                    ...  
8593                   1.0  
8595                   1.0  
8596                   1.0  
8597                   1.0  
8599                   1.0  

[5394 rows x 4 columns]

In [26]:
play_dataframe = pd.DataFrame(df[df["activity"] == "play"])
play_dataframe

user_id                        game_name activity  \
1     218323237                           Dota 2     play   
3     218323237                      Tomb Raider     play   
5     218323237                             TERA     play   
7     126340495                           Dota 2     play   
9     126340495  Counter-Strike Global Offensive     play   
...         ...                              ...      ...   
8590  138590829                           Dota 2     play   
8592  190460990                           Dota 2     play   
8594    7519923                   Counter-Strike     play   
8598  103963922                  Team Fortress 2     play   
8600  300991661                           Dota 2     play   

      hours_played_if_play  
1                     14.8  
3                      8.8  
5                      0.9  
7                   1784.0  
9                    626.0  
...                    ...  
8590                1026.0  
8592                   1.5  
8594                  35.0  
8598                   0.4  
8600                   0.3  

[3197 rows x 4 columns]

### 6. Percentage of Game Purchases that are Actually Played

What percentage of all game purchases in this dataset are actually played by the player that purchased them?

*You can assume that a game and user id that has a 'play' row ALSO MUST have a 'purchase' row (but not the other way around).*

**For full credit, use automated Python code to find your answer. (Ie. don't put something like '30/50', where you found the 30 and 50 from the output of other code.)**

In [28]:
purchase_count = purchase_dataframe["activity"].count() 
play_count = play_dataframe["activity"].count()
percentage_played = (play_count/purchase_count) * 100
percentage_played

59.26955876900259

### 7. Top Game Purchaser

Next, we'd like to learn more about the people who *purchase* the most amount of games.

#### 7.1. Most Amount of Games Purchased

What is the user id of the player who has purchased the most amount of games in this dataset, and how many games have they purchased?

In [287]:
df_purchase_grouped = pd.DataFrame(purchase_dataframe["user_id"].value_counts())
df_purchase_grouped.head()

count
user_id         
12610800     292
27279261     185
43913966     172
134322141    170
51822361     168

As we can see from the head, the user_id with the most purchased is '12610800' with a count of 292 games purchased.

#### 7.2. How many games has this user actually played?

What percentage of the purchased games did this player from 7.1 actually play?

In [290]:
df_12610800 = df[df["user_id"] == 12610800]
df_12610800

user_id                                      game_name  activity  \
917   12610800                                  Path of Exile  purchase   
918   12610800                                  Path of Exile      play   
919   12610800                                         Arma 3  purchase   
920   12610800                                         Arma 3      play   
921   12610800                                   PlanetSide 2  purchase   
...        ...                                            ...       ...   
1369  12610800            Vampire The Masquerade - Bloodlines  purchase   
1370  12610800                               Vertiginous Golf  purchase   
1371  12610800          Viscera Cleanup Detail Shadow Warrior  purchase   
1372  12610800                                         VVVVVV  purchase   
1373  12610800  Warlock - Master of the Arcane Powerful Lords  purchase   

      hours_played_if_play  
917                    1.0  
918                  329.0  
919                    1.0  
920                  325.0  
921                    1.0  
...                    ...  
1369                   1.0  
1370                   1.0  
1371                   1.0  
1372                   1.0  
1373                   1.0  

[457 rows x 4 columns]

In [294]:
df_12610800.shape[0]

457

In [292]:
purchase_12610800 = df_12610800[df_12610800["activity"] == "purchase"]
purchase_12610800

user_id                                      game_name  activity  \
917   12610800                                  Path of Exile  purchase   
919   12610800                                         Arma 3  purchase   
921   12610800                                   PlanetSide 2  purchase   
923   12610800                     Sid Meier's Civilization V  purchase   
925   12610800                                           DayZ  purchase   
...        ...                                            ...       ...   
1369  12610800            Vampire The Masquerade - Bloodlines  purchase   
1370  12610800                               Vertiginous Golf  purchase   
1371  12610800          Viscera Cleanup Detail Shadow Warrior  purchase   
1372  12610800                                         VVVVVV  purchase   
1373  12610800  Warlock - Master of the Arcane Powerful Lords  purchase   

      hours_played_if_play  
917                    1.0  
919                    1.0  
921                    1.0  
923                    1.0  
925                    1.0  
...                    ...  
1369                   1.0  
1370                   1.0  
1371                   1.0  
1372                   1.0  
1373                   1.0  

[292 rows x 4 columns]

In [293]:
purchase_count = purchase_12610800.shape[0]
purchase_count

292

In [295]:
play_12610800 = df_12610800[df_12610800["activity"] == "play"]
play_12610800

user_id                       game_name activity  hours_played_if_play
918   12610800                   Path of Exile     play                 329.0
920   12610800                          Arma 3     play                 325.0
922   12610800                    PlanetSide 2     play                 190.0
924   12610800      Sid Meier's Civilization V     play                 178.0
926   12610800                            DayZ     play                 171.0
...        ...                             ...      ...                   ...
1238  12610800                Unity of Command     play                   0.1
1240  12610800       Avadon The Black Fortress     play                   0.1
1242  12610800               Hack, Slash, Loot     play                   0.1
1244  12610800  Medal of Honor(TM) Multiplayer     play                   0.1
1246  12610800                    Rogue Legacy     play                   0.1

[165 rows x 4 columns]

In [296]:
play_count = play_12610800.shape[0]
play_count

165

In [302]:
percentage_12610800 = play_count / purchase_count * 100
percentage_12610800

56.5068493150685

## 8. Examining Some Game "All-Star" Players

Next, we'd like to examine the play vs. purchase behavior of a few games' "all star players". We will consider a game's "all star player" to be the one who has played the game the longest in this dataset.

### 8.1. Each Game's Highest Play Time

In a single dataframe below, display the highest amount of time that each game was played.

For full credit, your final resulting dataframe should:
1. Be comprised of *two columns*
 * game_name
 * hours_played_if_play
2. And should have an index going from 0,1,2,... etc.

In [304]:
df_group_by_game_hours = pd.DataFrame(df.groupby("game_name")["hours_played_if_play"].max())
df_group_by_game_hours.reset_index(inplace=True)
df_group_by_game_hours

game_name  hours_played_if_play
0                   10 Second Ninja                   1.0
1                 100% Orange Juice                   1.0
2                          404Sight                   1.0
3                     7 Days to Die                 336.0
4                           8BitBoy                  24.0
...                             ...                   ...
1564   realMyst Masterpiece Edition                   7.5
1565  resident evil 4 / biohazard 4                   3.8
1566                   sZone-Online                   1.0
1567      the static speaks my name                   1.0
1568                      theHunter                  95.0

[1569 rows x 2 columns]

### 8.2. Ranking the Games by Highest Maximum Game Play Time

Sort your final resulting dataframe from 8.1 by the number of hours that the game was played by the all-star player *in descending order*.

Overwrite your dataframe, so it remains sorted in this order.

In [306]:
df_group_by_game_hours = df_group_by_game_hours.sort_values("hours_played_if_play", ascending=False)
df_group_by_game_hours

game_name  hours_played_if_play
411                              Dota 2                5608.0
1156         Sid Meier's Civilization V                5002.0
112   Baldur's Gate II Enhanced Edition                3898.0
550               Football Manager 2013                3374.0
111      Baldur's Gate Enhanced Edition                2630.0
...                                 ...                   ...
340                Dead Island Epidemic                   1.0
342                         Dead Pixels                   1.0
344        Dead Rising 2 Off the Record                   1.0
346                  Dead Rising 3 DLC1                   1.0
784                        METAL SLUG 3                   1.0

[1569 rows x 2 columns]

### 8.3. Top 2 Games with the Most Dedicated "All-Star Player"

Which two games in this dataset had the most "dedicated" "all star players"? Or in other words, which two games were played the longest amount of time by a particular player?

In [329]:
df_group_by_game_hours.nlargest(2,"hours_played_if_play")

game_name  hours_played_if_play
411                       Dota 2                5608.0
1156  Sid Meier's Civilization V                5002.0

### 8.4. Identifying these "All Star Players"

For these two games that you selected in 8.3 above, what is the user id of their respective "all star player"?

In [338]:
two_largest = df.nlargest(2,"hours_played_if_play")
two_largest

user_id                   game_name activity  hours_played_if_play
2306   67694595                      Dota 2     play                5608.0
8122  128065805  Sid Meier's Civilization V     play                5002.0

As we can see above, the two user ids of the all star player for each of the two longest played games are "67694595" for the game Dota 2 which was played for 5608 hours and "128065805" for Sid Meier's Civilization V which was played for 5002 hours.

## 9. Learning about the All-Star Players

Finally, let's learn more about the play vs. purchase behavior of these two "all star" players that we identified in #8. Specifically for each of these two player user ids that you identified in #8, what percentage of games that they bought did they actually play?

In [339]:
df_67694595 = df[df["user_id"] == 67694595]
df_67694595

user_id                                      game_name  activity  \
2305  67694595                                         Dota 2  purchase   
2306  67694595                                         Dota 2      play   
2307  67694595                Counter-Strike Global Offensive  purchase   
2308  67694595                Counter-Strike Global Offensive      play   
2309  67694595                                  Path of Exile  purchase   
...        ...                                            ...       ...   
2455  67694595  Warlock - Master of the Arcane Powerful Lords  purchase   
2456  67694595                                    Worms Blast  purchase   
2457  67694595                               Worms Crazy Golf  purchase   
2458  67694595                                  Worms Pinball  purchase   
2459  67694595                          Worms Ultimate Mayhem  purchase   

      hours_played_if_play  
2305                   1.0  
2306                5608.0  
2307                   1.0  
2308                 463.0  
2309                   1.0  
...                    ...  
2455                   1.0  
2456                   1.0  
2457                   1.0  
2458                   1.0  
2459                   1.0  

[155 rows x 4 columns]

In [341]:
purchase_67694595 = df_67694595[df_67694595["activity"] == "purchase"]
purchase_67694595

user_id                                      game_name  activity  \
2305  67694595                                         Dota 2  purchase   
2307  67694595                Counter-Strike Global Offensive  purchase   
2309  67694595                                  Path of Exile  purchase   
2311  67694595                     The Elder Scrolls V Skyrim  purchase   
2313  67694595                                       PAYDAY 2  purchase   
...        ...                                            ...       ...   
2455  67694595  Warlock - Master of the Arcane Powerful Lords  purchase   
2456  67694595                                    Worms Blast  purchase   
2457  67694595                               Worms Crazy Golf  purchase   
2458  67694595                                  Worms Pinball  purchase   
2459  67694595                          Worms Ultimate Mayhem  purchase   

      hours_played_if_play  
2305                   1.0  
2307                   1.0  
2309                   1.0  
2311                   1.0  
2313                   1.0  
...                    ...  
2455                   1.0  
2456                   1.0  
2457                   1.0  
2458                   1.0  
2459                   1.0  

[95 rows x 4 columns]

In [349]:
purchase_count_67694595 = purchase_67694595.shape[0]
purchase_count_67694595

95

In [343]:
play_67694595 = df_67694595[df_67694595["activity"] == "play"]
play_67694595

user_id                                          game_name activity  \
2306  67694595                                             Dota 2     play   
2308  67694595                    Counter-Strike Global Offensive     play   
2310  67694595                                      Path of Exile     play   
2312  67694595                         The Elder Scrolls V Skyrim     play   
2314  67694595                                           PAYDAY 2     play   
2316  67694595                                      Torchlight II     play   
2318  67694595                       The Witcher Enhanced Edition     play   
2320  67694595                                        Titan Quest     play   
2322  67694595                  Dark Souls Prepare to Die Edition     play   
2324  67694595                        Titan Quest Immortal Throne     play   
2326  67694595                                        Dying Light     play   
2328  67694595                                      DARK SOULS II     play   
2330  67694595                                         Torchlight     play   
2332  67694595               Call of Duty Black Ops - Multiplayer     play   
2334  67694595                                            Magicka     play   
2336  67694595                                    Game Dev Tycoon     play   
2338  67694595                                    Cities Skylines     play   
2340  67694595                                    Team Fortress 2     play   
2342  67694595        The Incredible Adventures of Van Helsing II     play   
2344  67694595                                           Portal 2     play   
2346  67694595                                      Borderlands 2     play   
2348  67694595                                   Metro Last Light     play   
2350  67694595                              FTL Faster Than Light     play   
2352  67694595                          Call of Juarez Gunslinger     play   
2354  67694595                         Sid Meier's Civilization V     play   
2356  67694595                                      Left 4 Dead 2     play   
2358  67694595                                       PlanetSide 2     play   
2360  67694595                             Call of Duty Black Ops     play   
2362  67694595                                      Mirror's Edge     play   
2364  67694595                          Chivalry Medieval Warfare     play   
2366  67694595                                           Warframe     play   
2368  67694595                   Star Wars The Force Unleashed II     play   
2370  67694595                      Middle-earth Shadow of Mordor     play   
2372  67694595                                Magicka Wizard Wars     play   
2374  67694595                          Super Monday Night Combat     play   
2376  67694595  Star Wars The Force Unleashed Ultimate Sith Ed...     play   
2378  67694595                                         Darksiders     play   
2380  67694595                               The Binding of Isaac     play   
2382  67694595                                   Worms Armageddon     play   
2384  67694595                                         Metro 2033     play   
2386  67694595                                            Bastion     play   
2388  67694595                                        Alien Swarm     play   
2390  67694595                                    Korwin The Game     play   
2392  67694595              Company of Heroes (New Steam Version)     play   
2394  67694595                                            Trine 2     play   
2396  67694595                  Burnout Paradise The Ultimate Box     play   
2398  67694595                          Heroes of Might & Magic V     play   
2400  67694595                                Assassin's Creed II     play   
2402  67694595                                              Trine     play   
2404  67694595                                           ArcheAge     play   
2406  67694595               

In [350]:
play_count_67694595 = play_67694595.shape[0]
play_count_67694595

60

In [352]:
percentage_67694595 = (play_count_67694595 / purchase_count_67694595) * 100
percentage_67694595

63.1578947368421

In [340]:
df_128065805 = df[df["user_id"] == 128065805]
df_128065805

user_id                     game_name  activity  hours_played_if_play
8121  128065805    Sid Meier's Civilization V  purchase                   1.0
8122  128065805    Sid Meier's Civilization V      play                5002.0
8123  128065805  Age of Empires II HD Edition  purchase                   1.0
8124  128065805  Age of Empires II HD Edition      play                 364.0

In [353]:
purchase_128065805 = df_128065805[df_128065805["activity"] == "purchase"]
purchase_128065805

user_id                     game_name  activity  hours_played_if_play
8121  128065805    Sid Meier's Civilization V  purchase                   1.0
8123  128065805  Age of Empires II HD Edition  purchase                   1.0

In [367]:
purchase_shape_128065805 = purchase_128065805.shape[0]
purchase_shape_128065805

2

In [364]:
play_128065805 = df_128065805[df_128065805["activity"] == "play"]
play_128065805

user_id                     game_name activity  hours_played_if_play
8122  128065805    Sid Meier's Civilization V     play                5002.0
8124  128065805  Age of Empires II HD Edition     play                 364.0

In [368]:
play_shape_128065805 = play_128065805.shape[0]
play_shape_128065805

2

In [369]:
percentage_128065805 = (play_shape_128065805 / purchase_shape_128065805) * 100
percentage_128065805

100.0